core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java

```diff
@@ -514,6 +544,185 @@ public XContentBuilder toInnerXContent(XContentBuilder builder, Params params) t
+                        while((token = parser.nextToken()) != XContentParser.Token.END_ARRAY) {
+                            values.add(parseValue(parser));
+                        }
+                        fields.put(fieldName, new InternalSearchHitField(fieldName, values));
+                    }
+                } else if (Fields._EXPLANATION.equals(currentFieldName)) {
+                    explanation = parseExplanation(parser);
+                } else if (Fields.INNER_HITS.equals(currentFieldName)) {
+                    innerHits = new HashMap<>();
+                    while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+                        // parse the key
+                        ensureExpectedToken(XContentParser.Token.FIELD_NAME, parser.currentToken(), parser::getTokenLocation);
+                        String name = parser.currentName();
+                        parser.nextToken();
+                        innerHits.put(name, InternalSearchHits.fromXContent(parser));
+                        parser.nextToken();
+                    }
+                } else {
+                    throwUnknownField(currentFieldName, parser.getTokenLocation());
+                }
+            } else if (token == XContentParser.Token.START_ARRAY) {
+                if (Fields.MATCHED_QUERIES.equals(currentFieldName)) {
+                    while((token = parser.nextToken()) != XContentParser.Token.END_ARRAY) {
+                        matchedQueries.add(parser.text());
+                    }
+                } else {
+                    throwUnknownField(currentFieldName, parser.getTokenLocation());
+                }
+            } else {
+                throw new ParsingException(parser.getTokenLocation(),
+                        "Failed to parse object: unsupported token found [" + token + "]");
+            }
+        }
+        InternalSearchHit internalSearchHit = new InternalSearchHit(-1, id, new Text(type), nestedIdentity, Collections.emptyMap());
+        internalSearchHit.index(index);
+        internalSearchHit.score(score);
+        internalSearchHit.version(version);
+        internalSearchHit.sortValues(sortValues);
+        internalSearchHit.highlightFields(highlightFields);
+        internalSearchHit.sourceRef(parsedSource);
+        internalSearchHit.explanation(explanation);
+        internalSearchHit.setInnerHits(innerHits);
+        if (matchedQueries.size() > 0) {
+            internalSearchHit.matchedQueries(matchedQueries.toArray(new String[matchedQueries.size()]));
+        }
+        if (shardId != null || nodeId != null) {
+            internalSearchHit.shard(new SearchShardTarget(nodeId, shardId));
+        }
+        internalSearchHit.fields(fields);
+        return internalSearchHit;
+    }
+
+    private static Explanation parseExplanation(XContentParser parser) throws IOException {
+        XContentParser.Token token;
+        float value = 0.0f;
```

Comment:
I think I will make this a Float so we can validate if both description and value have been parsed before we return the Explanation, otherwise we throw.